<div class="alert alert-block alert-info"><b></b>
<h1><center> <font color='black'> Homework 05</font></center></h1>
<h2><center> <font color='black'> A/B Testing & Uplift modelling</font></center></h2>    
<h2><center> <font color='black'> MTAT.03.319 - Business Data Analytics</font></center></h2>
<h2><center> <font color='black'> University of Tartu - Spring 2023</font></center></h2>
</div>

# Homework instructions

- Please provide the names and student IDs of the team-members (Maximum 2 person) in the field "Team mates" below. If you are not working in a team please insert only your name and student ID.

- Please provide code where ever applicable.

- The accepted submission format is .ipynb file. If you are sharing Colab link, make sure that the privacy settings for the file is public so we can access your code. 

- The submission will automatically close on <font color='red'>**30 April at 23:59**</font>, so please make sure to submit before the deadline. 

- ONLY one of the teammates should submit the homework and in the submission description the other person's Name and Student ID must be entered. We will grade the homework and the marks and feedback is applied for both the team members. So please communicate with your team member about marks and feedback if you are submit the homework.

- If a question is not clear, please ask us in Moodle ONLY. 

- After you have finished solving the Homework, please restart the Kernel and run all the cells to check if there is any persisting issues. 

- Plagiarism is <font color='red'>**PROHIBITED**</font>. Any form of plagiarism will be dealt according to the university policy (https://ut.ee/en/content/academic-fraud).

- <font color='red'>**DO NOT CHANGE THE TEMPLATE**</font>

- <font color='red'>**Restart the Kernel and Run all the cells once again after you are done.**</font>
This will ensure that all the cells run without error. You will find an option in the top menu bar under Kernel tab. 

**<h2><font color='red'>Team mates:</font></h2>**


<font color='red'>Name: </font> Samir Musali &emsp; <font color='red'>Student ID: </font> C29810

### The homework is divided into two sections and the points are distributed as below:
<pre>
- A/B Testing               -> 5.5 points
- Uplift modeling           -> 5.5 points
________________________________________________
Total                       -> 11.0 points
</pre>

# 1. A/B Testing (5.5 points)


**1.1 Use `AB_clicks.csv` data and find whether Learn, Help, and Services versions of the page compared to the Interact have significantly more (or less) clicks. Justify the choice of the performed tests and interpret the result of the test. (1.5 points)** 

In [1]:
import pandas as pd
from scipy.stats import f_oneway

# Load the data into a pandas DataFrame, skipping any lines with errors
data = pd.read_csv('AB_clicks.csv', delimiter='\t')

# Extract the click data for Interact version
interact_clicks = data[data['Version'] == 'Interact']['No_clicks']

# Extract the click data for Learn version and perform the ANOVA test
learn_clicks = data[data['Version'] == 'Learn']['No_clicks']
f_statistic_learn, p_value_learn = f_oneway(interact_clicks, learn_clicks)
print('=====')
print('F statistic for Learn:', f_statistic_learn)
print('p-value for Learn:', p_value_learn)
print('=====')

# Extract the click data for Help version and perform the ANOVA test
help_clicks = data[data['Version'] == 'Help']['No_clicks']
f_statistic_help, p_value_help = f_oneway(interact_clicks, help_clicks)
print('=====')
print('F statistic for Help:', f_statistic_help)
print('p-value for Help:', p_value_help)
print('=====')

# Extract the click data for Services version and perform the ANOVA test
services_clicks = data[data['Version'] == 'Services']['No_clicks']
f_statistic_services, p_value_services = f_oneway(interact_clicks, services_clicks)
print('=====')
print('F statistic for Services:', f_statistic_services)
print('p-value for Services:', p_value_services)
print('=====')

=====
F statistic for Learn: 1.0480354270900327
p-value for Learn: 0.3078759180594482
=====
=====
F statistic for Help: 0.7133607164859326
p-value for Help: 0.39995700324796557
=====
=====
F statistic for Services: 1.0426424596354409
p-value for Services: 0.30926185388653943
=====



**1.2 Use two different methods to test the normality of two different versions of your choice (Connect, Help, Services,..) (2.0 points)**





In [2]:
from scipy.stats import shapiro, anderson

# Extract the click data for Help and Learn versions
help_clicks = data[data['Version'] == 'Help']['No_clicks']
learn_clicks = data[data['Version'] == 'Learn']['No_clicks']

# Test the normality of the Help clicks using Shapiro-Wilk test
shapiro_stat_help, shapiro_p_help = shapiro(help_clicks)
print('=====')
print('Shapiro-Wilk test for Help clicks:')
print('Statistic:', shapiro_stat_help)
print('p-value:', shapiro_p_help)
print('=====')

# Test the normality of the Learn clicks using Anderson-Darling test
anderson_stat_learn, anderson_critvals_learn, anderson_siglvl_learn = anderson(learn_clicks)
print('=====')
print('Anderson-Darling test for Learn clicks:')
print('Statistic:', anderson_stat_learn)
print('Critical values:', anderson_critvals_learn)
print('Significance levels:', anderson_siglvl_learn)
print('=====')

=====
Shapiro-Wilk test for Help clicks:
Statistic: 0.32242465019226074
p-value: 9.140807719304565e-15
=====
=====
Anderson-Darling test for Learn clicks:
Statistic: 15.69964300122372
Critical values: [0.544 0.62  0.744 0.868 1.032]
Significance levels: [15.  10.   5.   2.5  1. ]
=====


**1.3 Read about Multiple comparison problem read ([wikipedia](https://en.wikipedia.org/wiki/Multiple_comparisons_problem), [An Overview of Methods to Address the Multiple Comparison Problem](https://towardsdatascience.com/an-overview-of-methods-to-address-the-multiple-comparison-problem-310427b3ba92) or other resources). What is the problem and when we need to do something about it (describe briefly)? (1 points)**

<font color='red'> **Answer:** The Multiple Comparison Problem (MCP) arises when multiple hypotheses are tested on the same data set, increasing the likelihood of making at least one type I error (rejecting a true null hypothesis). In other words, as the number of comparisons increases, the possibility of observing at least one significant result by chance alone also increases.
When performing multiple hypothesis tests on the same data set, we must do something about the Multiple Comparison Problem (MCP). One common approach to address this problem is to adjust the significance level using a correction method such as the Bonferroni correction, which involves dividing the original significance level by the number of comparisons made. Another approach is to use a False Discovery Rate (FDR) correction, which controls the expected proportion of false discoveries among all discoveries.
Ignoring the Multiple Comparison Problem (MCP) can result in false positive results, which can have serious consequences, especially in fields such as medicine and finance, where decisions based on statistical significance can impact people's lives and financial outcomes.

**1.4 Load the dataset classifier_results.csv. The dataset contains the following fields:** 

**Dataset** - 200 datasets retreived from [OpenML](https://www.openml.org/search?type=data) dataset directory.

**Classifier** - 9 classifiers from scikit-learn

**Accuracy** - The accuracy achieved after applying the classifier to the dataset


**Is there a significant statistical difference between these classifiers ? If you were asked to continue your analysis with only a limited number of classifiers how many would you pick and why ? (1.0 points)** (Hint:You can use the orange package that was introduced in the practice session) 

In [3]:
import pandas as pd
from scipy.stats import f_oneway

# Load the dataset
data = pd.read_csv('classifier_results.csv')

# Extract the accuracy data for each classifier
classifiers = data['Classifier'].unique()
accuracies = []
for classifier in classifiers:
    classifier_accuracies = data[data['Classifier'] == classifier]['Accuracy']
    accuracies.append(classifier_accuracies)

# Perform one-way ANOVA
f_statistic, p_value = f_oneway(*accuracies)

# Print the ANOVA results
print('F statistic:', f_statistic)
print('p-value:', p_value)

F statistic: 0.6010799163946169
p-value: 0.7759837327632423


<font color='red'> **Answer:** The ANOVA results show that the p-value is much more significant than the commonly used significance level of 0.05. This suggests that there is not enough evidence to reject the null hypothesis, which states that there is no significant difference between the means of the accuracy scores for the different classifiers. If we were asked to continue the analysis with only a limited number of classifiers, we could select the top performers based on their mean accuracy scores or some other performance metric. Alternatively, we could use a post hoc test like Tukey's HSD (Honestly Significant Difference) to determine which classifiers are significantly different. This would allow us to identify specific pairs of classifiers that are very different while controlling for the overall Type I error rate.

# 2. Uplift modelling (5.5 points)
For this task we are going to use email_marketing.csv from [MineThatData](https://blog.minethatdata.com/2008/03/minethatdata-e-mail-analytics-and-data.html). The data was recorded from a two-weeks email campaign that included 64000 customers. Out of these customers:
1. 1/3 were randomly chosen to receive an e-mail campaign featuring Mens merchandise.
2. 1/3 were randomly chosen to receive an e-mail campaign featuring Womens merchandise.
3. 1/3 were randomly chosen to not receive an e-mail campaign.

The features in this dataset include:

- Recency: Months since last purchase.
- History: Actual dollar value spent in the past year.
- Mens: 1/0 indicator, 1 = customer purchased Mens merchandise in the past year.
- Womens: 1/0 indicator, 1 = customer purchased Womens merchandise in the past year.
- Zip_Code: Classifies zip code as Urban, Suburban, or Rural.
- Newbie: 1/0 indicator, 1 = New customer in the past twelve months.
- Channel: Describes the channels the customer purchased from in the past year.
- Segment: e-mail campaign  type that the customer received
- Visit: 1/0 indicator, 1 = Customer visited website in the following two weeks.

The last variable is realted to the evaluation if the campaign was effective or not.

- Conversion: 1/0 indicator, 1 = Customer purchased merchandise in the following two weeks.


**2.0 Load the dataset and provide description. (0.25 points)**

In [4]:
import pandas as pd

# Load the dataset into a pandas DataFrame
df = pd.read_csv('email_marketing.csv')

# Print the shape of the dataset (number of rows and columns)
print("Shape of dataset:", df.shape)

# Print the first 5 rows of the dataset
print("\nFirst 5 rows of dataset:\n", df.head())

# Print summary statistics for the numerical columns
print("\nSummary statistics for numerical columns:\n", df.describe())

# Print the data types of the columns
print("\nData types of columns:\n", df.dtypes)

Shape of dataset: (64000, 12)

First 5 rows of dataset:
    recency history_segment  history  mens  womens   zip_code  newbie channel  \
0       10  2) $100 - $200   142.44     1       0  Surburban       0   Phone   
1        6  3) $200 - $350   329.08     1       1      Rural       1     Web   
2        7  2) $100 - $200   180.65     0       1  Surburban       1     Web   
3        9  5) $500 - $750   675.83     1       0      Rural       1     Web   
4        2    1) $0 - $100    45.34     1       0      Urban       0     Web   

         segment  visit  conversion  spend  
0  Womens E-Mail      0           0    0.0  
1      No E-Mail      0           0    0.0  
2  Womens E-Mail      0           0    0.0  
3    Mens E-Mail      0           0    0.0  
4  Womens E-Mail      0           0    0.0  

Summary statistics for numerical columns:
             recency       history          mens        womens        newbie  \
count  64000.000000  64000.000000  64000.000000  64000.000000  64000.

**2.1 Use the function calc_uplift() from the practice session to calculate the current uplift of the campaign. Remeber to modify the function according to the need. (0.5 points)**

In [5]:
def calc_uplift(df):
    """
    Calculate the uplift generated by the campaign.
    
    Parameters:
    - df (pandas DataFrame): The dataset containing the campaign data.
    
    Returns:
    - uplift (float): The uplift generated by the campaign.
    """
    
    # Calculate the conversion rate for each group
    control_conv_rate = df[df['segment'] == 'No E-Mail']['conversion'].mean()
    mens_conv_rate = df[df['segment'] == 'Mens E-Mail']['conversion'].mean()
    womens_conv_rate = df[df['segment'] == 'Womens E-Mail']['conversion'].mean()
    
    # Calculate the overall conversion rate
    overall_conv_rate = df['conversion'].mean()
    
    # Calculate the uplift for each group
    control_uplift = mens_uplift = womens_uplift = 0
    if overall_conv_rate > 0:
        control_uplift = (mens_conv_rate + womens_conv_rate - 2 * overall_conv_rate) / overall_conv_rate
        mens_uplift = (mens_conv_rate - overall_conv_rate) / overall_conv_rate
        womens_uplift = (womens_conv_rate - overall_conv_rate) / overall_conv_rate
    
    # Calculate the overall uplift
    uplift = (df[df['segment'] == 'Mens E-Mail']['conversion'].sum() + df[df['segment'] == 'Womens E-Mail']['conversion'].sum()) / df['conversion'].sum() - 1
    
    # Print the results
    print('Control conversion rate:', control_conv_rate)
    print('Mens conversion rate:', mens_conv_rate)
    print('Womens conversion rate:', womens_conv_rate)
    print('Control uplift:', control_uplift)
    print('Mens uplift:', mens_uplift)
    print('Womens uplift:', womens_uplift)
    print('Overall uplift:', uplift)
    
    return uplift

uplift = calc_uplift(df)
print('Current uplift:', uplift)

Control conversion rate: 0.005726086548390125
Mens conversion rate: 0.01253109306800582
Womens conversion rate: 0.008837144059475383
Control uplift: 0.3660331767453238
Mens uplift: 0.387525876042167
Womens uplift: -0.02149269929684335
Overall uplift: -0.2110726643598616
Current uplift: -0.2110726643598616


**2.2 Create a new column named target_class. Split the customers in 4 categories:**

1- Treatment Responders 

2- Treatment Non-Responders (sleeping dogs)

3- Control Responders (sure things)

4-Control Non-Responders(lost causes).

**Label encode these categories and store the result for each customer in the column target_class (0.5 points)**

In [39]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Create a copy of the original dataframe
df_encoded = df.copy()

# Encode the non-numeric columns using one-hot encoding
non_numeric_cols = ['history_segment', 'zip_code', 'channel']
ohe = OneHotEncoder()
for col in non_numeric_cols:
    encoded_col = ohe.fit_transform(df_encoded[col].values.reshape(-1,1)).toarray()
    n_cols = encoded_col.shape[1]
    for i in range(n_cols):
        df_encoded[f"{col}_{i}"] = encoded_col[:, i]
    df_encoded.drop(col, axis=1, inplace=True)

# Create the target_class column
df_encoded['target_class'] = ''

# Define the boolean indexing conditions for each category
treatment_responders = (df_encoded['segment'].isin(['Mens E-Mail', 'Womens E-Mail'])) & (df_encoded['conversion'] == 1)
treatment_non_responders = (df_encoded['segment'].isin(['Mens E-Mail', 'Womens E-Mail'])) & (df_encoded['conversion'] == 0)
control_responders = (df_encoded['segment'] == 'No E-Mail') & (df_encoded['conversion'] == 1)
control_non_responders = (df_encoded['segment'] == 'No E-Mail') & (df_encoded['conversion'] == 0)

# Fit and transform the categories to their encoded labels
le = LabelEncoder()
categories = ['Treatment Responders', 'Treatment Non-Responders', 'Control Responders', 'Control Non-Responders']
le.fit(categories)

# Label encode the categories and store the result in the target_class column
df_encoded.loc[treatment_responders, 'target_class'] = le.transform(['Treatment Responders'])[0]
df_encoded.loc[treatment_non_responders, 'target_class'] = le.transform(['Treatment Non-Responders'])[0]
df_encoded.loc[control_responders, 'target_class'] = le.transform(['Control Responders'])[0]
df_encoded.loc[control_non_responders, 'target_class'] = le.transform(['Control Non-Responders'])[0]

# Print the unique values in the target_class column
print('Unique values in target_class:', df_encoded['target_class'].unique())

Unique values in target_class: [2 0 3 1]


**2.3 Apply the necessary preprocessing to the data and split it into train and test set, using 80/20 ratio. Remember that you have to drop the columns that define the label and perform preprocessing. Build a model with your classifier of choice that will predict probabilities for the categories mentioned in 2.2. (1.5 points)**

In [41]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Convert the target_class column to integer type
df_encoded['target_class'] = df_encoded['target_class'].astype(int)

# Drop columns that define the label
X = df_encoded.drop(['conversion', 'segment', 'target_class'], axis=1)
y = df_encoded['target_class']

# Perform preprocessing
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

from sklearn.ensemble import RandomForestClassifier

# Instantiate the classifier
rf = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the model on the training data
rf.fit(X_train, y_train)

# Predict probabilities for each class on the test data
y_prob = rf.predict_proba(X_test)

# Print the probabilities for each class
print('Probabilities for each class:')
for i, label in enumerate(le.classes_):
    print(label, y_prob[:, i])

Probabilities for each class:
Control Non-Responders [0.03 0.11 0.18 ... 0.05 0.09 0.5 ]
Control Responders [0.   0.   0.   ... 0.   0.22 0.  ]
Treatment Non-Responders [0.97 0.89 0.82 ... 0.95 0.02 0.5 ]
Treatment Responders [0.   0.   0.   ... 0.   0.67 0.  ]


**2.4 Calculate the uplift score using probabilities for all customers and store the results to a new column in the dataframe (0.5 points)**

In [45]:
# Calculate probabilities for each class on the entire data
y_prob_all = rf.predict_proba(X)

# Calculate expected outcome for each customer in treatment group
df_encoded['treatment_prob'] = y_prob_all[:, 2] + y_prob_all[:, 3]

# Calculate expected outcome for each customer in control group
df_encoded['control_prob'] = y_prob_all[:, 1] + y_prob_all[:, 0]

# Calculate uplift score for each customer
df_encoded['uplift_score'] = df_encoded['treatment_prob'] - df_encoded['control_prob']

# Visualize the uplift_score
print(df_encoded['uplift_score'].head())
print('=====')
print(df_encoded['uplift_score'].describe())

0    0.76
1   -0.76
2    0.84
3    0.96
4    0.72
Name: uplift_score, dtype: float64
=====
count    64000.000000
mean         0.334997
std          0.581210
min         -1.000000
25%         -0.240000
50%          0.520000
75%          0.840000
max          1.000000
Name: uplift_score, dtype: float64


**2.5  Select the customers with High Uplift Score as below:**
    
**High Uplift Score:** Customers have uplift score > 3rd quantile

**a. Calculate the ratio of the customers targeted with Womens e-mail that have high uplift score over the total customers who received emails featuring Womens merchandise. (0.5 points)**

**b. Calculate the ratio of the customers targeted with Mens e-mail that have high uplift score over the total customers who received emails featuring Mens merchandise. (0.5 points)**


**c.Which one of the campaigns would you say is more successful ? (0.5 points)**


In [46]:
# Calculate the 3rd quartile of uplift scores
uplift_q3 = df_encoded['uplift_score'].quantile(0.75)

# Select customers with high uplift score
df_high_uplift = df_encoded[df_encoded['uplift_score'] > uplift_q3]

# Calculate the ratio of customers with high uplift score for each campaign
womens_total = df_encoded[df_encoded['segment'] == 'Womens E-Mail'].shape[0]
womens_high_uplift = df_high_uplift[df_high_uplift['segment'] == 'Womens E-Mail'].shape[0]
womens_high_uplift_ratio = womens_high_uplift / womens_total

mens_total = df_encoded[df_encoded['segment'] == 'Mens E-Mail'].shape[0]
mens_high_uplift = df_high_uplift[df_high_uplift['segment'] == 'Mens E-Mail'].shape[0]
mens_high_uplift_ratio = mens_high_uplift / mens_total

print('Ratio of customers with high uplift score for Womens E-Mail:', womens_high_uplift_ratio)
print('Ratio of customers with high uplift score for Mens E-Mail:', mens_high_uplift_ratio)

Ratio of customers with high uplift score for Womens E-Mail: 0.34605133959882173
Ratio of customers with high uplift score for Mens E-Mail: 0.3521847280236542


<font color='red'> **Answer C:** Based on the ratios calculated above, the uplift scores for both campaigns are very similar. The ratio of customers with high uplift scores is slightly higher for the Mens E-Mail campaign. However, based on this information alone, we cannot definitively say which campaign is more successful. Success can be defined in different ways, depending on the business objectives. If the campaign's primary goal is to maximize the number of customers with high uplift scores, then the Mens E-Mail campaign is more successful. On the other hand, if the goal is to maximize the overall revenue, we need to consider the conversion rates and average order value of the customers targeted in each campaign. Therefore, to determine which campaign is more successful, we need to analyze the conversion rates and average order value of the customers targeted in each campaign. We can then compare these metrics between the two campaigns to determine which campaign is more effective in generating revenue for the business.

**2.6 For customers in the segment named "Womens e-mail" with high uplift score calculate conversion uplift and revenue uplift. Compare it with the benchmark from excercise 2.1 and draw conclusions. (0.75 points)**

In [49]:
def calc_uplift(df):
    """
    Calculate the uplift generated by the campaign.

    Parameters:
    - df (pandas DataFrame): The dataset containing the campaign data.

    Returns:
    - uplift (float): The uplift generated by the campaign.
    """

    # Calculate the conversion rate for each group
    control_conv_rate = df[df['segment'] == 'No E-Mail']['conversion'].mean()
    mens_conv_rate = df[df['segment'] == 'Mens E-Mail']['conversion'].mean()
    womens_conv_rate = df[df['segment'] == 'Womens E-Mail']['conversion'].mean()

    # Calculate the overall conversion rate
    overall_conv_rate = df['conversion'].mean()

    # Calculate the uplift for each group
    control_uplift = mens_uplift = womens_uplift = 0
    if overall_conv_rate > 0:
        control_uplift = (mens_conv_rate + womens_conv_rate - 2 * overall_conv_rate) / overall_conv_rate
        mens_uplift = (mens_conv_rate - overall_conv_rate) / overall_conv_rate
        womens_uplift = (womens_conv_rate - overall_conv_rate) / overall_conv_rate

    # Calculate the overall uplift
    uplift = (df[df['segment'] == 'Mens E-Mail']['conversion'].sum() + df[df['segment'] == 'Womens E-Mail']['conversion'].sum()) / df['conversion'].sum() - 1

    # Calculate the uplift for women with high uplift score
    df_womens_high_uplift = df[(df['segment'] == 'Womens E-Mail') & (df['uplift_score'] > df['uplift_score'].quantile(0.75))]
    womens_high_uplift_conv_rate = df_womens_high_uplift['conversion'].mean()
    womens_high_uplift_uplift = (womens_high_uplift_conv_rate - overall_conv_rate) / overall_conv_rate
    womens_high_uplift_rev = womens_high_uplift_uplift * df_womens_high_uplift['spend'].sum()

    # Print the results
    print('Control conversion rate:', control_conv_rate)
    print('Mens conversion rate:', mens_conv_rate)
    print('Womens conversion rate:', womens_conv_rate)
    print('Control uplift:', control_uplift)
    print('Mens uplift:', mens_uplift)
    print('Womens uplift:', womens_uplift)
    print('Overall uplift:', uplift)
    print('Womens high uplift conversion rate:', womens_high_uplift_conv_rate)
    print('Womens high uplift uplift:', womens_high_uplift_uplift)
    print('Womens high uplift revenue:', womens_high_uplift_rev)

    return

print('Overall campaign uplift:')
calc_uplift(df_encoded)

print('\nUplift for Womens E-Mail with High Uplift Score:')
calc_uplift(df_high_uplift[df_high_uplift['segment'] == 'Womens E-Mail'])

Overall campaign uplift:
Control conversion rate: 0.005726086548390125
Mens conversion rate: 0.01253109306800582
Womens conversion rate: 0.008837144059475383
Control uplift: 0.3660331767453238
Mens uplift: 0.387525876042167
Womens uplift: -0.02149269929684335
Overall uplift: -0.2110726643598616
Womens high uplift conversion rate: 0.011079583840021619
Womens high uplift uplift: 0.22680513107505823
Womens high uplift revenue: 2192.1441694823816

Uplift for Womens E-Mail with High Uplift Score:
Control conversion rate: nan
Mens conversion rate: nan
Womens conversion rate: 0.011079583840021619
Control uplift: nan
Mens uplift: nan
Womens uplift: 0.0
Overall uplift: 0.0
Womens high uplift conversion rate: 0.004140786749482402
Womens high uplift uplift: -0.6262687471595213
Womens high uplift revenue: -83.87617330707468


<font color='red'> **Answer:** Looking at the results, we can conclude that sending a Women's e-mail to customers with high uplift score yields positive conversion and revenue uplifts. For the entire dataset, the uplift rates for the control and men's groups are higher than the women's group, meaning that sending no email or a men's email yields better results than a women's email. The overall campaign has a negative uplift, which means that sending the emails did not result in an overall increase in conversion rate. However, for customers in the Women's e-mail segment with high uplift scores, the conversion rate is significantly higher than the overall conversion rate, resulting in a positive uplift of 22.68%. Sending a Women's e-mail to customers with high uplift scores considerably increased this segment's conversion rate. The revenue uplift for this segment is also positive, resulting in a revenue increase of 2192.14. Therefore, sending a Women's e-mail to customers with high uplift scores is more successful than not sending an email or a Men's e-mail for this segment.

## How long did it take you to solve the homework?

* Please answer as precisely as you can. It does not affect your points or grade in any way. It is okay, if it took 0.5 hours or 24 hours. The collected information will be used to improve future homeworks.

<font color='red'> **Answer: 16 hours**</font>


## What is the level of difficulty for this homework?
you can put only number between $0:10$ ($0:$ easy, $10:$ difficult)

<font color='red'> **Answer: 9**</font>